responses/tests/test_responses.py

```diff
@@ -2035,6 +2036,88 @@ def run():
     assert_reset()
 
 
+def test_response_and_requests_mock_calls_are_equal():
+    @responses.activate
+    def run():
+        rsp = responses.add(responses.GET, "http://www.example.com")
+        rsp2 = responses.add(responses.GET, "http://www.example.com/1")
+
+        requests.get("http://www.example.com")
+        requests.get("http://www.example.com/1")
+
+        assert len(responses.calls) == 2
+        assert rsp.call_count == 1
+        assert rsp.calls[0] is responses.calls[0]
+        assert rsp2.call_count == 1
+        assert rsp2.calls[0] is responses.calls[1]
+
+    run()
+    assert_reset()
+
+
+def test_response_call_request():
+    @responses.activate
+    def run():
+        rsp = responses.add(responses.GET, "http://www.example.com")
+        rsp2 = responses.add(
+            responses.PUT, "http://www.foo.bar/42/", json={"id": 42, "name": "Bazz"}
+        )
+
+        requests.get("http://www.example.com")
+        requests.get("http://www.example.com?hello=world")
+        requests.put(
+            "http://www.foo.bar/42/",
+            json={"name": "Bazz"},
+        )
+
+        assert rsp.call_count == 2
+        request = rsp.calls[0].request
+        assert request.url == "http://www.example.com/"
+        assert request.method == "GET"
+        request = rsp.calls[1].request
+        assert request.url == "http://www.example.com/?hello=world"
+        assert request.method == "GET"
+        assert rsp2.call_count == 1
+        request = rsp2.calls[0].request
+        assert request.url == "http://www.foo.bar/42/"
+        assert request.method == "PUT"
+        request_payload = json.loads(request.body)
```

Comment:
what exactly do we test here ?

why do we need to test that request body could be converted?

Reply:
(I changed the test a little)

Here we validate that request `rsp2_request1 = rsp2.calls[0].request` obtained from the mock's call contains expected payload `request_payload = {"name": "Bazz"}`, that was used on the put request.

> request body could be converted?

`PreparedRequest` contains only encoded data in the `body` field, so we decode it to compare with the original data - `request_payload`.

Reply:
> `PreparedRequest` contains only encoded data in the `body` field, so we decode it to compare with the original data - `request_payload`.

correct, but do we need to check it at all as part of `responses` ?

we attach `request` object as is without modification 

https://github.com/getsentry/responses/blob/5c6a3b13bbefc2d4cae0cbd1e9177ab702e15e76/responses/__init__.py#L1041

and in the case of this PR we can skip this check since it does not add a value.

although, you can fire a parallel PR that can add a separate test to validate that we have all the required data is attached to a request
we still will need to discuss if that is required or not but this we can do in a separate thread

Reply:
@beliaev-maksim and what about this question?

Reply:
@zeezdev sorry, missed this one in a long thread. Yes.
And if for some reason we have missed the tests that ensure the quality of `Call` objects. Then I would recommend to create separate tests for those and open it in a separate PR.